torchsnapshot/storage_plugins/fsspec.py

```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+import asyncio
+import io
+import os
+
+import fsspec
+
+from torchsnapshot.io_types import ReadIO, StoragePlugin, WriteIO
+
+__all__ = ["FSSpecStoragePlugin"]
+
+
+class FSSpecStoragePlugin(StoragePlugin):
+    def __init__(self, root: str) -> None:
+        root_items = root.split("://")
+        if len(root_items) != 2:
+            raise ValueError("only protocol://path is supported by fsspec plugin")
+        protocol, self.root = root_items
+        if not protocol.startswith("fsspec-"):
+            raise ValueError(
+                f"Invalid protocol: {protocol}, Only fsspec-* protocols are supported"
+            )
+        self.fs = fsspec.filesystem(protocol=protocol.removeprefix("fsspec-"))
+        self._session = None
+        self._lock = asyncio.Lock()
+
+    async def _init_session(self) -> None:
+        async with self._lock:
+            if self._session is None:
+                self._session = await self.fs.set_session()
+
+    async def write(self, write_io: WriteIO) -> None:
+        await self._init_session()
+        path = os.path.join(self.root, write_io.path)
+        await self.fs._pipe_file(path, bytes(write_io.buf))
+
+    async def read(self, read_io: ReadIO) -> None:
+        await self._init_session()
+        path = os.path.join(self.root, read_io.path)
+        result = await self.fs._cat_file(path)
```

Comment:
Are `_pipe_file`, `_cat_file`, and `_rm_file` async only APIs? IIUC, not all fsspec plugins have async support right?

Reply:
yes, It may differ according to the implement of fsspec storage plugin. But I invest some storage plugin of fsspec such as s3, gcsfs and azure blob. At least _pipe, _cat, _rm async method is supported.

all 3rd storage plugin for fsspec is listed here:
https://github.com/fsspec

Btw, the s3fs plugin doesn't support the async _open, so I use _pipe and _cat as a replace implement.

Reply:
Thank you for your due diligence! It is very informative.

I think we can start with the assumption that these methods are present in most fsspec plugins. Later we can add a thread pool-based fall back.